### PR TITLE
WIP Add Traverser for BinaryBackend index-based traversal

### PR DESCRIPTION
**Pull request by elbow-jason**
This PR is a work-in-progress.

This PR explores the viability of using an Enumerable to traverse the indices needed to transform binaries in many BinaryBackend operations via Traverser and AggIter.

This PR would allow #206 to be more easily achieved. This PR would also greatly simplify #182.

Pros:

- Most of the code of `Traverser` is a port of `BinaryBackend.aggregate_axes` and friends so it is familiar.
- `Traverser` would eliminate the need for intermediate binaries during aggregation (means fewer binary heap allocations).
- `Traverser` and `AggIter` are easily usable with the Enum module.
- `Traverser` and `AggIter` make the implementation of dot trivial (see tests).
- `Traverser` is responsive to reversal.
- `Traverser` is responsive to transpose.
- `Traverser` works with an empty axes list.

 Cons:
 
 - `Traverser` currently builds two list of indices for use in traversal. The lists are `offsets` and `readers` which are the expanded indices generated from `weighted_traverse`. 
- Who knows if it works correctly!?

@josevalim @seanmor5 Thoughts?

### CODE REVIEW

File: nx/lib/nx/binary_backend/traverser.ex [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+      offsets = Enum.map(offsets_template, fn o -> o + cycle_offset end)
+      %Traverser{trav | cycle: cycle, offsets: offsets, readers: readers_template}
+    end
+  end
+
+  defp next_item([]) do
+    :done
+  end
+
+  defp next_item(start..stop) when start == stop do
```
Comment (elbow-jason) [inline nx/lib/nx/binary_backend/traverser.ex:169]: The ranges here are relic of some experimentation. Their use might still be viable.

**Comment by elbow-jason** [comment]
Should `AggIter` be called `ViewIter`?

**Comment by josevalim** [comment]
Hi @elbow-jason! Thanks for looking into this, this direction looks exciting!

I think one of the things you will want to do is to benchmark the current implementation with the previous one. The code as today is not the best thing to work with but it is the most performant I could come up with. The backend is already slow and making it slower can be prohibitive. I believe some patterns, like the use of protocols, will have to revisited if we want to squeeze maximum performance. Instead you should likely provide high-level APIs on the traverser such as reduce, flat_map, etc that are invoked directly and skip the protocol. So I think dot would end-up looking like:

    def dot(%{data: data1, view: view1}, axes1, %{data: data2, view: view2}, axes2) do
      view1 = View.to_traverse(view1, axes1)
      view2 = View.to_traverse(view2, axes2)

      # traverse2 to traverse two at the same time. The view knows what it
      # needs to traverse and on which order, as in your current code today.
      View.traverse2(data1, view1, data2, view2, fn ..., ... -> do_dot(...) end)
    end

Also try to keep the data structures as simple as possible. Remember that when working on a tensor of 1000x1000, we would be calling something like `reduce` 1 million times to read sequential bytes in a bytearray, so any kind of reading of a maps, updating of maps, will be an incredible amount of overhead.

I hope this provides some guidance! This is definitely the direction that we want! To recap, I think we should keep the view and data as two distinct things, which you are already doing. Working offsets is a great idea, because it help us simplify things and it should not affect performance. And then simplify the view construct as much possible. :)

**Comment by elbow-jason** [comment]
I am excited too! Good point about the performance. I'll need to benchmark sooner rather than later.

**Comment by elbow-jason** [comment]
A simple bench:
```
➜  nx git:(jg-traverser) ✗ mix run bench/dot.exs
Generated nx app
md benches {100, 100}
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
Number of Available Cores: 8
Available memory: 15.59 GB
Elixir 1.11.2
Erlang 23.1.5

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
parallel: 1
inputs: none specified
Estimated total run time: 56 s

Benchmarking dot4_bin md_f32...
Benchmarking dot4_bin md_f64...
Benchmarking dot4_trav md_f32...
Benchmarking dot4_trav md_f64...

Name                       ips        average  deviation         median         99th %
dot4_trav md_f32          4.80      208.50 ms     ±3.10%      207.07 ms      223.05 ms
dot4_trav md_f64          4.61      216.79 ms     ±4.02%      214.66 ms      255.97 ms
dot4_bin md_f32           4.51      221.95 ms     ±2.78%      221.29 ms      241.84 ms
dot4_bin md_f64           4.47      223.90 ms     ±4.00%      221.58 ms      266.53 ms

Comparison: 
dot4_trav md_f32          4.80
dot4_trav md_f64          4.61 - 1.04x slower +8.28 ms
dot4_bin md_f32           4.51 - 1.06x slower +13.45 ms
dot4_bin md_f64           4.47 - 1.07x slower +15.39 ms

Memory usage statistics:

Name                     average  deviation         median         99th %
dot4_trav md_f32       252.55 MB     ±0.00%      252.55 MB      252.55 MB
dot4_trav md_f64       252.55 MB     ±0.00%      252.55 MB      252.55 MB
dot4_bin md_f32        331.81 MB     ±0.00%      331.81 MB      331.83 MB
dot4_bin md_f64        331.79 MB     ±0.01%      331.80 MB      331.82 MB

Comparison: 
dot4_trav md_f32       252.55 MB
dot4_trav md_f64       252.55 MB - 1.00x memory usage +0 MB
dot4_bin md_f32        331.81 MB - 1.31x memory usage +79.26 MB
dot4_bin md_f64        331.79 MB - 1.31x memory usage +79.24 MB
sm benches {4, 4}
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
Number of Available Cores: 8
Available memory: 15.59 GB
Elixir 1.11.2
Erlang 23.1.5

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
parallel: 1
inputs: none specified
Estimated total run time: 1.40 min

Benchmarking dot4_bin sm_f32...
Benchmarking dot4_bin sm_f64...
Benchmarking dot4_bin sm_u8...
Benchmarking dot4_trav sm_f32...
Benchmarking dot4_trav sm_f64...
Benchmarking dot4_trav sm_u8...

Name                       ips        average  deviation         median         99th %
dot4_trav sm_u8        60.56 K       16.51 μs    ±76.45%       14.09 μs       51.32 μs
dot4_bin sm_u8         59.49 K       16.81 μs    ±61.50%       14.45 μs       47.10 μs
dot4_trav sm_f32       55.50 K       18.02 μs    ±36.50%       15.94 μs       45.97 μs
dot4_bin sm_f32        53.08 K       18.84 μs    ±60.19%       16.64 μs       48.96 μs
dot4_trav sm_f64       52.92 K       18.90 μs    ±60.43%       16.12 μs       54.75 μs
dot4_bin sm_f64        49.04 K       20.39 μs    ±58.78%       17.55 μs       63.53 μs

Comparison: 
dot4_trav sm_u8        60.56 K
dot4_bin sm_u8         59.49 K - 1.02x slower +0.30 μs
dot4_trav sm_f32       55.50 K - 1.09x slower +1.51 μs
dot4_bin sm_f32        53.08 K - 1.14x slower +2.33 μs
dot4_trav sm_f64       52.92 K - 1.14x slower +2.39 μs
dot4_bin sm_f64        49.04 K - 1.23x slower +3.88 μs

Memory usage statistics:

Name                Memory usage
dot4_trav sm_u8         17.59 KB
dot4_bin sm_u8          27.02 KB - 1.54x memory usage +9.43 KB
dot4_trav sm_f32        19.59 KB - 1.11x memory usage +2 KB
dot4_bin sm_f32         29.16 KB - 1.66x memory usage +11.57 KB
dot4_trav sm_f64        19.59 KB - 1.11x memory usage +2 KB
dot4_bin sm_f64         29.30 KB - 1.67x memory usage +11.72 KB

**All measurements for memory usage were the same**
lg rect benches dot({400, 4}, {4, 400})
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
Number of Available Cores: 8
Available memory: 15.59 GB
Elixir 1.11.2
Erlang 23.1.5

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
parallel: 1
inputs: none specified
Estimated total run time: 28 s

Benchmarking dot4_bin lg_rect_f64...
Benchmarking dot4_trav lg_rect_f64...

Name                            ips        average  deviation         median         99th %
dot4_bin lg_rect_f64           5.98      167.28 ms     ±4.03%      164.94 ms      189.66 ms
dot4_trav lg_rect_f64          5.87      170.39 ms     ±6.48%      167.84 ms      226.26 ms

Comparison: 
dot4_bin lg_rect_f64           5.98
dot4_trav lg_rect_f64          5.87 - 1.02x slower +3.11 ms

Memory usage statistics:

Name                     Memory usage
dot4_bin lg_rect_f64        221.47 MB
dot4_trav lg_rect_f64       173.40 MB - 0.78x memory usage -48.07120 MB

**All measurements for memory usage were the same**
sm rect benches dot({4, 400}, {400, 4})
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
Number of Available Cores: 8
Available memory: 15.59 GB
Elixir 1.11.2
Erlang 23.1.5

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
parallel: 1
inputs: none specified
Estimated total run time: 28 s

Benchmarking dot4_bin sm_rect_f64...
Benchmarking dot4_trav sm_rect_f64...

Name                            ips        average  deviation         median         99th %
dot4_trav sm_rect_f64        696.06        1.44 ms    ±13.66%        1.38 ms        2.23 ms
dot4_bin sm_rect_f64         600.70        1.66 ms    ±15.43%        1.58 ms        2.63 ms

Comparison: 
dot4_trav sm_rect_f64        696.06
dot4_bin sm_rect_f64         600.70 - 1.16x slower +0.23 ms

Memory usage statistics:

Name                     Memory usage
dot4_trav sm_rect_f64         1.64 MB
dot4_bin sm_rect_f64          2.59 MB - 1.58x memory usage +0.95 MB

**All measurements for memory usage were the same**
```

**Comment by elbow-jason** [comment]
Bench results for new sort vs old sort:

```
➜  nx git:(jg-traverser) mix run bench/sort.exs
md benches axis 0 {100, 100}
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
Number of Available Cores: 8
Available memory: 15.59 GB
Elixir 1.11.2
Erlang 23.1.5

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
parallel: 1
inputs: none specified
Estimated total run time: 1.87 min

Benchmarking new sort md_f32 axis: 0...
Benchmarking new sort md_f64 axis: 0...
Benchmarking new sort md_s64 axis: 0...
Benchmarking new sort md_u8 axis: 0...
Benchmarking old sort md_f32 axis: 0...
Benchmarking old sort md_f64 axis: 0...
Benchmarking old sort md_s64 axis: 0...
Benchmarking old sort md_u8 axis: 0...

Name                              ips        average  deviation         median         99th %
new sort md_u8 axis: 0         208.11        4.81 ms     ±5.78%        4.76 ms        5.50 ms
new sort md_f64 axis: 0        198.73        5.03 ms     ±5.71%        4.97 ms        5.69 ms
new sort md_s64 axis: 0        197.25        5.07 ms     ±5.71%        5.03 ms        5.76 ms
new sort md_f32 axis: 0        191.32        5.23 ms    ±11.40%        5.13 ms        7.16 ms
old sort md_u8 axis: 0         162.63        6.15 ms     ±9.62%        6.00 ms        7.35 ms
old sort md_f64 axis: 0        158.65        6.30 ms     ±5.00%        6.26 ms        7.13 ms
old sort md_s64 axis: 0        157.97        6.33 ms     ±5.83%        6.27 ms        7.34 ms
old sort md_f32 axis: 0        155.55        6.43 ms     ±7.58%        6.42 ms        7.46 ms

Comparison: 
new sort md_u8 axis: 0         208.11
new sort md_f64 axis: 0        198.73 - 1.05x slower +0.23 ms
new sort md_s64 axis: 0        197.25 - 1.06x slower +0.26 ms
new sort md_f32 axis: 0        191.32 - 1.09x slower +0.42 ms
old sort md_u8 axis: 0         162.63 - 1.28x slower +1.34 ms
old sort md_f64 axis: 0        158.65 - 1.31x slower +1.50 ms
old sort md_s64 axis: 0        157.97 - 1.32x slower +1.53 ms
old sort md_f32 axis: 0        155.55 - 1.34x slower +1.62 ms

Memory usage statistics:

Name                       Memory usage
new sort md_u8 axis: 0          3.84 MB
new sort md_f64 axis: 0         3.99 MB - 1.04x memory usage +0.144 MB
new sort md_s64 axis: 0         3.84 MB - 1.00x memory usage +0 MB
new sort md_f32 axis: 0         3.99 MB - 1.04x memory usage +0.144 MB
old sort md_u8 axis: 0          5.13 MB - 1.34x memory usage +1.29 MB
old sort md_f64 axis: 0         5.10 MB - 1.33x memory usage +1.26 MB
old sort md_s64 axis: 0         5.13 MB - 1.34x memory usage +1.29 MB
old sort md_f32 axis: 0         5.14 MB - 1.34x memory usage +1.29 MB

**All measurements for memory usage were the same**
md benches axis 1 {100, 100}
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
Number of Available Cores: 8
Available memory: 15.59 GB
Elixir 1.11.2
Erlang 23.1.5

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
parallel: 1
inputs: none specified
Estimated total run time: 1.87 min

Benchmarking new sort md_f32 axis: 1...
Benchmarking new sort md_f64 axis: 1...
Benchmarking new sort md_s64 axis: 1...
Benchmarking new sort md_u8 axis: 1...
Benchmarking old sort md_f32 axis: 1...
Benchmarking old sort md_f64 axis: 1...
Benchmarking old sort md_s64 axis: 1...
Benchmarking old sort md_u8 axis: 1...

Name                              ips        average  deviation         median         99th %
new sort md_u8 axis: 1         295.53        3.38 ms    ±10.90%        3.44 ms        4.18 ms
old sort md_u8 axis: 1         293.25        3.41 ms    ±12.42%        3.47 ms        4.21 ms
old sort md_f32 axis: 1        282.62        3.54 ms    ±11.72%        3.59 ms        4.32 ms
new sort md_f32 axis: 1        277.48        3.60 ms    ±12.27%        3.67 ms        4.51 ms
old sort md_f64 axis: 1        277.48        3.60 ms    ±12.61%        3.67 ms        4.51 ms
new sort md_f64 axis: 1        276.35        3.62 ms    ±10.67%        3.67 ms        4.37 ms
new sort md_s64 axis: 1        274.27        3.65 ms    ±12.15%        3.70 ms        4.47 ms
old sort md_s64 axis: 1        271.14        3.69 ms    ±12.33%        3.74 ms        4.60 ms

Comparison: 
new sort md_u8 axis: 1         295.53
old sort md_u8 axis: 1         293.25 - 1.01x slower +0.0263 ms
old sort md_f32 axis: 1        282.62 - 1.05x slower +0.155 ms
new sort md_f32 axis: 1        277.48 - 1.07x slower +0.22 ms
old sort md_f64 axis: 1        277.48 - 1.07x slower +0.22 ms
new sort md_f64 axis: 1        276.35 - 1.07x slower +0.23 ms
new sort md_s64 axis: 1        274.27 - 1.08x slower +0.26 ms
old sort md_s64 axis: 1        271.14 - 1.09x slower +0.30 ms

Memory usage statistics:

Name                       Memory usage
new sort md_u8 axis: 1          2.76 MB
old sort md_u8 axis: 1          2.76 MB - 1.00x memory usage +0 MB
old sort md_f32 axis: 1         2.91 MB - 1.05x memory usage +0.152 MB
new sort md_f32 axis: 1         2.91 MB - 1.05x memory usage +0.152 MB
old sort md_f64 axis: 1         2.91 MB - 1.05x memory usage +0.152 MB
new sort md_f64 axis: 1         2.91 MB - 1.05x memory usage +0.152 MB
new sort md_s64 axis: 1         2.76 MB - 1.00x memory usage +0 MB
old sort md_s64 axis: 1         2.76 MB - 1.00x memory usage +0 MB

**All measurements for memory usage were the same**
sm benches axis 0 {100, 100}
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
Number of Available Cores: 8
Available memory: 15.59 GB
Elixir 1.11.2
Erlang 23.1.5

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
parallel: 1
inputs: none specified
Estimated total run time: 1.40 min

Benchmarking new sort sm_f32 axis: 0...
Benchmarking new sort sm_f64 axis: 0...
Benchmarking new sort sm_u8 axis: 0...
Benchmarking old sort sm_f32 axis: 0...
Benchmarking old sort sm_f64 axis: 0...
Benchmarking old sort sm_u8 axis: 0...

Name                              ips        average  deviation         median         99th %
new sort sm_u8 axis: 0        64.37 K       15.53 μs    ±69.08%       14.29 μs       40.70 μs
new sort sm_f32 axis: 0       62.36 K       16.03 μs   ±105.04%       14.61 μs       42.66 μs
new sort sm_f64 axis: 0       60.88 K       16.43 μs    ±70.02%       14.95 μs       46.72 μs
old sort sm_u8 axis: 0        43.71 K       22.88 μs    ±26.77%       21.59 μs       50.75 μs
old sort sm_f32 axis: 0       43.05 K       23.23 μs    ±42.52%       21.65 μs       54.66 μs
old sort sm_f64 axis: 0       41.25 K       24.24 μs    ±58.00%       22.30 μs       56.37 μs

Comparison: 
new sort sm_u8 axis: 0        64.37 K
new sort sm_f32 axis: 0       62.36 K - 1.03x slower +0.50 μs
new sort sm_f64 axis: 0       60.88 K - 1.06x slower +0.89 μs
old sort sm_u8 axis: 0        43.71 K - 1.47x slower +7.34 μs
old sort sm_f32 axis: 0       43.05 K - 1.50x slower +7.70 μs
old sort sm_f64 axis: 0       41.25 K - 1.56x slower +8.71 μs

Memory usage statistics:

Name                       Memory usage
new sort sm_u8 axis: 0         10.40 KB
new sort sm_f32 axis: 0        10.84 KB - 1.04x memory usage +0.44 KB
new sort sm_f64 axis: 0        10.77 KB - 1.04x memory usage +0.38 KB
old sort sm_u8 axis: 0         13.60 KB - 1.31x memory usage +3.20 KB
old sort sm_f32 axis: 0        13.83 KB - 1.33x memory usage +3.43 KB
old sort sm_f64 axis: 0        14.09 KB - 1.35x memory usage +3.69 KB

**All measurements for memory usage were the same**
sm benches axis 1 {100, 100}
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
Number of Available Cores: 8
Available memory: 15.59 GB
Elixir 1.11.2
Erlang 23.1.5

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
parallel: 1
inputs: none specified
Estimated total run time: 1.40 min

Benchmarking new sort sm_f32 axis: 1...
Benchmarking new sort sm_f64 axis: 1...
Benchmarking new sort sm_u8 axis: 1...
Benchmarking old sort sm_f32 axis: 1...
Benchmarking old sort sm_f64 axis: 1...
Benchmarking old sort sm_u8 axis: 1...

Name                              ips        average  deviation         median         99th %
new sort sm_u8 axis: 1       138.06 K        7.24 μs   ±305.24%        6.33 μs       21.28 μs
old sort sm_u8 axis: 1       137.51 K        7.27 μs   ±269.79%        6.32 μs       21.47 μs
new sort sm_f32 axis: 1      128.89 K        7.76 μs   ±239.51%        6.82 μs       22.12 μs
old sort sm_f32 axis: 1      128.69 K        7.77 μs   ±236.94%        6.81 μs       22.31 μs
old sort sm_f64 axis: 1      125.83 K        7.95 μs   ±232.60%        6.96 μs       22.80 μs
new sort sm_f64 axis: 1      125.32 K        7.98 μs   ±253.46%        6.98 μs       22.58 μs

Comparison: 
new sort sm_u8 axis: 1       138.06 K
old sort sm_u8 axis: 1       137.51 K - 1.00x slower +0.0292 μs
new sort sm_f32 axis: 1      128.89 K - 1.07x slower +0.52 μs
old sort sm_f32 axis: 1      128.69 K - 1.07x slower +0.53 μs
old sort sm_f64 axis: 1      125.83 K - 1.10x slower +0.70 μs
new sort sm_f64 axis: 1      125.32 K - 1.10x slower +0.74 μs

Memory usage statistics:

Name                       Memory usage
new sort sm_u8 axis: 1          5.94 KB
old sort sm_u8 axis: 1          5.94 KB - 1.00x memory usage +0 KB
new sort sm_f32 axis: 1         6.31 KB - 1.06x memory usage +0.38 KB
old sort sm_f32 axis: 1         6.31 KB - 1.06x memory usage +0.38 KB
old sort sm_f64 axis: 1         6.28 KB - 1.06x memory usage +0.34 KB
new sort sm_f64 axis: 1         6.28 KB - 1.06x memory usage +0.34 KB

**All measurements for memory usage were the same**
```

**Comment by elbow-jason** [comment]
The benchmarks are looking pretty promising.

I think we could get even faster by "projecting" one traversal through another, but it might get complicated.

@josevalim Take a look at the changes to [sort](https://github.com/elixir-nx/nx/pull/322/files#diff-15095458a1b2fb983edb557f9d4b0664e25425d06d053665f48867c8df015307R1819-R1833) and let me know what you think of the approach.

**Comment by josevalim** [comment]
Hi @elbow-jason, this is looking great!

I think we should start working towards incorporating those changes. One of the things I would like to propose is to reduce the amount of layers. Today we have a BinaryBackend -> TensorView -> View and that all adds complexity. My suggestion is incorporate the view data directly into the binary backend struct.

To be more precise: I would send a PR that adds the views fields to the BinaryBackend and change both from_binary/to_binary to consider this field may be set. As a first step, we will implement transpose on top of views. This should be a small and granular step that will allow us to convert the rest step by step. WDYT?

**Comment by elbow-jason** [comment]
@josevalim Sounds good to me. I'll move the view into the BinaryBackend and condense the code into fewer layers. I have been struggling with separation-of-concerns boundaries and code layout.

**Comment by josevalim** [comment]
Closing this as it has gone stale for now. Thank you @elbow-jason! :heart:

**Pull request closed without merging**